src/main/java/org/opensearch/flowframework/transport/DeprovisionWorkflowTransportAction.java

```diff
@@ -0,0 +1,337 @@
+                getResourcesAndExecute(request.getWorkflowId(), provisionProcessSequence, listener);
+            }, exception -> {
+                if (exception instanceof FlowFrameworkException) {
+                    logger.error("Workflow validation failed for workflow : " + workflowId);
+                    listener.onFailure(exception);
+                } else {
+                    logger.error("Failed to retrieve template from global context.", exception);
+                    listener.onFailure(new FlowFrameworkException(exception.getMessage(), ExceptionsHelper.status(exception)));
+                }
+            }));
+        } catch (Exception e) {
+            String message = "Failed to retrieve template from global context.";
+            logger.error(message, e);
+            listener.onFailure(new FlowFrameworkException(message, ExceptionsHelper.status(e)));
+        }
+    }
+
+    private void getResourcesAndExecute(
+        String workflowId,
+        List<ProcessNode> provisionProcessSequence,
+        ActionListener<WorkflowResponse> listener
+    ) {
+        GetWorkflowStateRequest getStateRequest = new GetWorkflowStateRequest(workflowId, true);
+        client.execute(GetWorkflowStateAction.INSTANCE, getStateRequest, ActionListener.wrap(response -> {
+            // Get a map of step id to created resources
+            final Map<String, ResourceCreated> resourceMap = response.getWorkflowState()
+                .resourcesCreated()
+                .stream()
+                .collect(Collectors.toMap(ResourceCreated::workflowStepId, Function.identity()));
+
+            // Now finally do the deprovision
+            executeDeprovisionSequence(workflowId, resourceMap, provisionProcessSequence, listener);
+        }, exception -> {
+            String message = "Failed to get workflow state for workflow " + workflowId;
+            logger.error(message, exception);
+            listener.onFailure(new FlowFrameworkException(message, ExceptionsHelper.status(exception)));
+        }));
+    }
+
+    private void executeDeprovisionSequence(
+        String workflowId,
+        Map<String, ResourceCreated> resourceMap,
+        List<ProcessNode> provisionProcessSequence,
+        ActionListener<WorkflowResponse> listener
+    ) {
+        // Create a list of ProcessNodes with the corresponding deprovision workflow steps
+        List<ProcessNode> deprovisionProcessSequence = provisionProcessSequence.stream()
+            // Only include nodes that created a resource
+            .filter(pn -> resourceMap.containsKey(pn.id()))
+            // Create a new ProcessNode with a deprovision step
+            .map(pn -> {
+                String stepName = pn.workflowStep().getName();
+                String deprovisionStep = getDeprovisionStepByWorkflowStep(stepName);
+                // Unimplemented steps presently return null, so skip
+                if (deprovisionStep == null) {
+                    return null;
```

Comment:
if we don't have an associated deprovision step, will it just skip this?

Reply:
That's correct, so for example create index won't delete index (yet).

Reply:
we can maybe use this concept for deciding not to deprovisiong some things on purpose? For example if i want to delete and deprovision everything but the index

Reply:
Seems a blunt instrument for that application.    I'll be thinking about update/fine-grained deprovisioning as more APIs become available  -- but this is the point in the code where that might be skipped.